server/app/mutations/domain_authorizations/authorize.rb

```diff
@@ -0,0 +1,77 @@
+require 'acme-client'
+require 'openssl'
+
+require_relative '../grid_certificates/common'
+require_relative '../../services/logging'
+
+module GridDomainAuthorizations
+  class Authorize < Mutations::Command
+    include GridCertificates::Common
+    include Logging
+
+    required do
+      model :grid, class: Grid
+      string :domain
+      string :authorization_type, in: ['dns-01', 'tls-sni-01'], default: 'dns-01'
+    end
+
+    optional do
+      string :linked_service
+    end
+
+    def validate
+      if self.authorization_type == 'tls-sni-01'
+        add_error(:linked_service, :missing, "Service link needs to be given for tls-sni-01 authorization type") unless self.linked_service
+      end
+      if self.linked_service
+        add_error(:linked_service, :not_found, "Linked service needs to point to existing service") unless @linked_service = resolve_service(self.grid, linked_service)
+      end
+    end
+
+    def execute
+      authorization = acme_client(self.grid).authorize(domain: self.domain)
+      challenge = nil
+      case self.authorization_type
+      when 'dns-01'
+        debug "creating dns-01 challenge"
+        challenge = authorization.dns01
+        challenge_opts = {
+          'record_name' => challenge.record_name,
+          'record_type' => challenge.record_type,
+          'record_content' => challenge.record_content
+        }
+      when 'tls-sni-01'
+        debug "creating tls-sni-01 challenge"
+        challenge = authorization.tls_sni01
```

Comment:
Hmm... what happens if the ACME server doesn't offer a `tls-sni-01` challenge? Like if using this with a wildcard domain, and it only offers `dns-01`?

Reply:
Currently wildcards are not supported at all. Support for those are coming early 2018 with a completely new ACME v2 API which probably means completely new API on acme client also. So as of now I don't see a case where this would fail like that.

Reply:
Yeah, that's just speculation, but I'm just thinking what this does if ACME doesn't offer such a challenge... is there some error that gets raised?

Reply:
doesn't seems so:
https://github.com/unixcharles/acme-client/blob/10f0f7ebf2b9f882a95409ee69c019923e4e4822/lib/acme/client/resources/authorization.rb#L30-L42

Reply:
Okay, so the `challenge = ` will be `nil`? Maybe also worth checking for that

Reply:
done.